core/domain/collection_services.py

```diff
@@ -687,6 +686,40 @@ def validate_exps_in_collection_are_public(collection):
                 'collection, exploration ID: %s' % exploration_id)
 
 
+def _get_sorted_nodes_in_dict(collection_nodes):
```

Comment:
@DubeySandeep -- I thought we agreed on something different here? What I recall was that the nodes should be sorted in the _frontend_ prior to the creator being allowed to click the Save or Publish button. When they get to the backend, the backend simply validates that the order of the nodes is valid, but does not sort them or change them in any way. If the order of the nodes is invalid, the backend raises an error and the creator has to fix it before saving (though this hopefully shouldn't happen in practice since the frontend should try to do error-checking there as well).

This code does not seem to do that.

Reply:
:( but lastly, you changed the entire thing and asked me to do so...
>For now, all frontend collections are currently linear. I suggest that just prior to saving to the backend, you look at the order of stuff in the collection and adjust the prereq/acquired skills accordingly to match.
>
>That way, you don't run the risk of showing validation errors that the creator cannot fix. And the stuff that's saved to the backend is always fully valid.

[Link to doc](https://docs.google.com/document/d/1NDvSHlvm94_OsFqyOwemgFM13SXqtlKzQbqxITK4wcE/edit?ts=5a36255d#)

Reply:
Er, my previous comment in this thread asks for a similar thing as the comment you quoted...

> I suggest that just prior to saving to the backend, you look at the order of stuff in the collection and adjust the prereq/acquired skills accordingly to match.

"Prior to saving in the backend" means doing the updates and checking in the frontend before sending the payload to the backend. In both cases the backend still just does the validation.

That said, you are correct that, in the frontend, the creator can't currently fix things, so auto-fixing it for them would be better. (In that respect, my comment in this thread was incorrect, sorry about that.)

However, the general point that all the editing of the collection dict should happen in the frontend, before save, still stands.